### PR TITLE
Replace challenge behavior with 1st class Forbid

### PR DESCRIPTION
**Pull request by HaoK**
AuthZ will be responsible for determining challenge/forbid

@Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Authentication.Core/AuthenticationService.cs [hunk 1]
```diff
@@ -100,11 +99,40 @@ public AuthenticationService(IAuthenticationSchemeProvider schemes, IAuthenticat
...
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <param name="scheme">The name of the authentication scheme.</param>
+        /// <param name="properties">The <see cref="AuthenticationProperties"/>.</param>
+        /// <returns>A task.</returns>
+        public virtual async Task ForbidAsync(HttpContext context, string scheme, AuthenticationProperties properties)
+        {
+            if (scheme == null)
+            {
+                // REVIEW: does it make sense to reuse the same default as challenge still?
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Core/AuthenticationService.cs]: Yes, defaulting to challenge is fine.

**Review by Tratcher — APPROVED** [review]
You'll need to update tests in HttpSysServer, IISIntegration, ServerTests, and Security?

File: src/Microsoft.AspNetCore.Authentication.Abstractions/ChallengeContext.cs [hunk 2]
```diff
@@ -26,20 +26,13 @@ public ChallengeContext(HttpContext httpContext, string authenticationScheme)
         /// <param name="httpContext">The context.</param>
         /// <param name="authenticationScheme">The name of the scheme.</param>
         /// <param name="properties">The properties.</param>
-        /// <param name="behavior">The challenge behavior.</param>
-        public ChallengeContext(HttpContext httpContext, string authenticationScheme, AuthenticationProperties properties, ChallengeBehavior behavior)
+        public ChallengeContext(HttpContext httpContext, string authenticationScheme, AuthenticationProperties properties)
             : base(httpContext, authenticationScheme, properties)
         {
             if (string.IsNullOrEmpty(authenticationScheme))
```
Comment (PinpointTownes) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/ChallengeContext.cs]: You should consider removing this, as it's a check already enforced by the base context class.

File: src/Microsoft.AspNetCore.Authentication.Abstractions/ForbidContext.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="httpContext">The context.</param>
+        /// <param name="authenticationScheme">The name of the scheme.</param>
+        /// <param name="properties">The properties.</param>
+        public ForbidContext(HttpContext httpContext, string authenticationScheme, AuthenticationProperties properties)
+            : base(httpContext, authenticationScheme, properties)
+        {
+            if (string.IsNullOrEmpty(authenticationScheme))
```
Comment (PinpointTownes) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/ForbidContext.cs]: Same remark :smile:

**Comment by PinpointTownes** [comment]
Finally!

**Comment by HaoK** [comment]
Updated to nuke context classes

File: src/Microsoft.AspNetCore.Authentication.Abstractions/BaseContext.cs [hunk 4]
```diff
@@ -1,49 +0,0 @@
...
-    /// <summary>
-    /// Base class used by other context classes.
-    /// </summary>
-    public abstract class BaseContext
-    {
-        /// <summary>
-        /// Constructor.
-        /// </summary>
-        /// <param name="context">The request context.</param>
-        protected BaseContext(HttpContext context)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/BaseContext.cs]: This was used by all the event contexts, are you moving it back to security?

**Review by Tratcher — APPROVED** [review]

File: src/Microsoft.AspNetCore.Authentication.Abstractions/BaseContext.cs (hunk 4 above)
Comment (HaoK) [inline src/Microsoft.AspNetCore.Authentication.Abstractions/BaseContext.cs]: Yeah back to security

**Comment by HaoK** [comment]
b1f92fb6bc42b2cdbb62bd930cedaebafece1e0a

**Pull request closed without merging**